pkg/operator/etcdcertcleaner/etcd_cert_cleaner_controller.go

```diff
@@ -0,0 +1,187 @@
+	tlshelpers.GetServingMetricsSecretNameForNode(""),
+}
+
+// EtcdCertCleanerController will clean up old and unused dynamically generated ("node-based") secrets.
+// This is to avoid filling up etcd in environments where the control plane is rotated/scaled up and down very often.
+type EtcdCertCleanerController struct {
+	eventRecorder      events.Recorder
+	kubeClient         kubernetes.Interface
+	operatorClient     v1helpers.StaticPodOperatorClient
+	masterNodeLister   corev1listers.NodeLister
+	masterNodeSelector labels.Selector
+	secretInformer     corev1informers.SecretInformer
+	secretLister       corev1listers.SecretLister
+	secretClient       corev1client.SecretsGetter
+	quorumChecker      ceohelpers.QuorumChecker
+}
+
+func NewEtcdCertCleanerController(
+	livenessChecker *health.MultiAlivenessChecker,
+	kubeClient kubernetes.Interface,
+	operatorClient v1helpers.StaticPodOperatorClient,
+	kubeInformers v1helpers.KubeInformersForNamespaces,
+	masterNodeLister corev1listers.NodeLister,
+	masterNodeSelector labels.Selector,
+	eventRecorder events.Recorder,
+) factory.Controller {
+	secretInformer := kubeInformers.InformersFor(operatorclient.TargetNamespace).Core().V1().Secrets()
+	secretLister := secretInformer.Lister()
+	secretClient := v1helpers.CachedSecretGetter(kubeClient.CoreV1(), kubeInformers)
+
+	c := &EtcdCertCleanerController{
+		eventRecorder:      eventRecorder,
+		kubeClient:         kubeClient,
+		operatorClient:     operatorClient,
+		masterNodeLister:   masterNodeLister,
+		masterNodeSelector: masterNodeSelector,
+		secretInformer:     secretInformer,
+		secretLister:       secretLister,
+		secretClient:       secretClient,
+	}
+
+	// we estimate the rate of control plane scales to maybe 1-2 a day, so running this once an hour with 6h of health threshold
+	// should be more than sufficient. Note that this controller does not sync on any informers.
+	syncer := health.NewCheckingSyncWrapper(c.sync, 6*time.Hour)
+	livenessChecker.Add("EtcdCertCleanerController", syncer)
+	return factory.New().ResyncEvery(1*time.Hour).WithSync(syncer.Sync).ToController("EtcdCertCleanerController", c.eventRecorder)
+}
+
+// sync will try to find unused cert secrets which are then marked for deletion, after three days the secrets are deleted.
+// This control loop does not implement "undelete". In case the secrets become managed by CertSignerController again, the
+// deletion label is getting removed automatically - in the worst case the cert is getting recreated entirely.
+func (c *EtcdCertCleanerController) sync(ctx context.Context, _ factory.SyncContext) error {
+	secrets, err := c.findUnusedNodeBasedSecrets(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, secretToDelete := range secrets {
+		klog.Infof("deleting secret [%s] detected unused for %s", secretToDelete.Name, DeletionGracePeriod)
+		err := c.secretClient.Secrets(operatorclient.TargetNamespace).Delete(ctx, secretToDelete.Name, v1.DeleteOptions{})
```

Comment:
The dynamic certs are all in the `openshift-etcd` target namespace so that makes sense. I was wondering if we could have unused certs in the other namespaces that we copy to e.g openshift-config. Don't think those are dynamic e.g client or ca-bundle, so no.

Reply:
Yep, good point. 

The cool feature of the secret/configmap sync is that it deletes when the origin resources disappears. Maybe you remember that this caused some issues in the nightlies the other day and TRT had to revert ;)

It's enough to clear the "source of truth", the rest follows along nicely.